ratis-metrics/src/main/java/org/apache/ratis/metrics/MetricsReporting.java

```diff
@@ -34,17 +34,41 @@ private MetricsReporting() {
   }
 
   public static Consumer<RatisMetricRegistry> consoleReporter(TimeDuration rate) {
-    return ratisMetricRegistry -> ConsoleReporter.forRegistry(ratisMetricRegistry.getDropWizardMetricRegistry())
-        .convertRatesTo(TimeUnit.SECONDS).convertDurationsTo(TimeUnit.MILLISECONDS).build()
-        .start(rate.getDuration(), rate.getUnit());
+    return ratisMetricRegistry -> {
+      ConsoleReporter reporter = ConsoleReporter.forRegistry(ratisMetricRegistry.getDropWizardMetricRegistry())
+          .convertRatesTo(TimeUnit.SECONDS).convertDurationsTo(TimeUnit.MILLISECONDS).build();
+      reporter.start(rate.getDuration(), rate.getUnit());
+      ratisMetricRegistry.setConsoleReporter(reporter);
+    };
+  }
+
+  public static Consumer<RatisMetricRegistry> stopConsoleReporter() {
```

Comment:
Do we need to add a stop method for every reporter? Maybe we can rely on the generic close() method implemented by all reports (Closeable) and they already stop() in the close() method?

Reply:
@avijayanhwx Yes, we can close() JmxReporter and ConsoleReporter. But for prometheus, it only has unregister, and also need a parameter i.e.collection. I think prometheus make the problem complicated.

![image](https://user-images.githubusercontent.com/51938049/80747058-66b76f00-8b55-11ea-8402-5f60ecb03ed2.png)

Reply:
@avijayanhwx For ConsoleReporter and JmxReporter we can call close. But for prometheus in ozone, there is only unregister() method. The following code copied from https://github.com/apache/hadoop-ozone/pull/885.
![image](https://user-images.githubusercontent.com/51938049/80791141-48885800-8bc3-11ea-9d7f-56076b4c3671.png)

Reply:
Thank you @runzhiwang. Can we use the close method now, given that you have removed the prometheus reporter registrations? Is that OK with you?

Reply:
@avijayanhwx Hi, thanks for review. Even though I have removed prometheus in ratis, but it exists in ozone. Such as the following code. ozone call `CollectorRegistry.defaultRegistry.register` to register registry in prometheus, and it also need call `CollectorRegistry.defaultRegistry.unregister` to unregister it. There is no close method for prometheus, and the unregister also need a parameter which differs from Reporter::close.

```
    MetricRegistries.global().addReporterRegistration(
        registry -> CollectorRegistry.defaultRegistry.register(
            new RatisDropwizardExports(
                registry.getDropWizardMetricRegistry())),
        registry -> CollectorRegistry.defaultRegistry.unregister(
            new RatisDropwizardExports(
                registry.getDropWizardMetricRegistry())));
```

Reply:
@runzhiwang Thanks, but IMHO, given that the JMX and console reporters' implementation are present in Ratis code itself, we can expose an APIs that look like enableJmxReporter(), enableConsoleReporters(), and the current addReporterRegistration API can be used for external reporters like Prometheus. This keeps a client code simple, but provides scope for adding custom reporters.

Reply:
@avijayanhwx Thanks for explanation. I agree with you. I have exposed enableJmxReporter(), enableConsoleReporters().

Reply:
Thanks @runzhiwang.